crates/tracing/src/layers.rs

```diff
@@ -0,0 +1,172 @@
+    /// * `S` - The type of subscriber that will use these layers.
+    ///
+    /// # Arguments
+    /// * `format` - The log message format.
+    /// * `directive` - Directive for the default logging level.
+    /// * `filter` - Additional filter directives as a string.
+    /// * `color` - Optional color configuration for the log messages.
+    ///
+    /// # Returns
+    /// An `eyre::Result<()>` indicating the success or failure of the operation.
+    pub(crate) fn stdout(
+        &mut self,
+        format: LogFormat,
+        directive: Directive,
+        filter: &str,
+        color: Option<String>,
+    ) -> eyre::Result<()> {
+        let filter = build_env_filter(Some(directive), filter)?;
+        let layer = format.apply(filter, color, None);
+        self.inner.push(layer.boxed());
+        Ok(())
+    }
+
+    /// Adds a file logging layer to the layers collection.
+    ///
+    /// # Arguments
+    /// * `format` - The format for log messages.
+    /// * `filter` - Additional filter directives as a string.
+    /// * `file_info` - Information about the log file including path and rotation strategy.
+    ///
+    /// # Returns
+    /// An `eyre::Result<FileWorkerGuard>` representing the file logging worker.
+    pub(crate) fn file(
+        &mut self,
+        format: LogFormat,
+        filter: String,
+        file_info: FileInfo,
+    ) -> eyre::Result<FileWorkerGuard> {
+        let log_dir = file_info.create_log_dir();
+        let (writer, guard) = file_info.create_log_writer(log_dir);
+
+        let file_filter = build_env_filter(None, &filter)?;
+        let layer = format.apply(file_filter, None, Some(writer));
+        self.inner.push(layer);
+        Ok(guard)
+    }
+}
+
+/// Holds configuration information for file logging.
+///
+/// Contains details about the log file's path, name, size, and rotation strategy.
+
+#[derive(Debug)]
+pub struct FileInfo {
+    dir: PathBuf,
+    file_name: String,
+    max_size_bytes: u64,
+    max_files: usize,
+}
+
+impl FileInfo {
+    /// Creates the log directory if it doesn't exist.
+    ///
+    /// # Returns
+    /// A reference to the path of the log directory.
+    fn create_log_dir(&self) -> &Path {
+        let log_dir: &Path = self.dir.as_ref();
+        if !log_dir.exists() {
+            std::fs::create_dir_all(log_dir).expect("Could not create log directory");
+        }
+        log_dir
+    }
+
+    /// Creates a non-blocking writer for the log file.
+    ///
+    /// # Arguments
+    /// * `log_dir` - Reference to the log directory path.
+    ///
+    /// # Returns
+    /// A tuple containing the non-blocking writer and its associated worker guard.
+    fn create_log_writer(
+        &self,
+        log_dir: &Path,
+    ) -> (tracing_appender::non_blocking::NonBlocking, WorkerGuard) {
+        let (writer, guard) = tracing_appender::non_blocking(
+            RollingFileAppender::new(
+                log_dir.join(AsRef::<Path>::as_ref(&self.file_name)),
```

Comment:
I think it should work with just a reference
```suggestion
                log_dir.join(&self.file_name),
```